core/storage/statistics/gae_models_test.py

```diff
@@ -43,16 +46,179 @@ def test_create_and_get_event_models(self):
+            submitted_answer_list)
+
+        model1 = stats_models.StateAnswersModel.get_master_model(
+            'exp_id', 1, 'state_name')
+
+        # Ensure we got the correct model.
+        self.assertEqual(model1.exploration_id, 'exp_id')
+        self.assertEqual(model1.exploration_version, 1)
+        self.assertEqual(model1.state_name, 'state_name')
+        self.assertEqual(model1.submitted_answer_list, submitted_answer_list)
+        self.assertEqual(model1.shard_count, 0)
+
+        # Use a smaller max answer list size so fewer answers are needed to
+        # exceed a shard. This will increase the 'shard_count'.
+        with self.swap(
+            stats_models.StateAnswersModel, '_MAX_ANSWER_LIST_BYTE_SIZE', 1):
+            stats_models.StateAnswersModel.insert_submitted_answers(
+                'exp_id', 1, 'state_name', 'interaction_id',
+                submitted_answer_list)
+
+            model1 = stats_models.StateAnswersModel.get_master_model(
+                'exp_id', 1, 'state_name')
+
+            self.assertEqual(model1.shard_count, 1)
+
+            stats_models.StateAnswersModel.insert_submitted_answers(
+                'exp_id', 1, 'state_name', 'interaction_id',
+                submitted_answer_list)
+
+            model1 = stats_models.StateAnswersModel.get_master_model(
+                'exp_id', 1, 'state_name')
+
+            self.assertEqual(model1.shard_count, 2)
+
+        # 'shard_count' will not increase as number of answers are less than
+        # the max answer list size.
+        stats_models.StateAnswersModel.insert_submitted_answers(
+            'exp_id', 1, 'state_name', 'interaction_id',
+            submitted_answer_list)
+
+        model1 = stats_models.StateAnswersModel.get_master_model(
+            'exp_id', 1, 'state_name')
+
+        self.assertEqual(model1.shard_count, 2)
+
+
+class StateAnswersCalcOutputModelTests(test_utils.GenericTestBase):
+
+    def test_create_and_get_models(self):
+
+        stats_models.StateAnswersCalcOutputModel.create_or_update(
+            'exp_id', '1', 'state_name', '', 'calculation_id', '', {})
+
+        model1 = stats_models.StateAnswersCalcOutputModel.get_model(
+            'exp_id', '1', 'state_name', 'calculation_id')
+
+        self.assertEqual(model1.exploration_id, 'exp_id')
+        self.assertEqual(model1.exploration_version, '1')
+        self.assertEqual(model1.state_name, 'state_name')
+        self.assertEqual(model1.calculation_id, 'calculation_id')
+
+        # Model with 'invalid_exp_id' does not exist.
+        model1 = stats_models.StateAnswersCalcOutputModel.get_model(
+            'invalid_exp_id', '1', 'state_name', 'calculation_id')
+
+        self.assertIsNone(model1)
+
+    def test_raise_exception_with_large_calculation_output(self):
+
+        observed_log_messages = []
+
+        def _mock_logging_function(msg, *_):
+            """Mocks logging.exception."""
+            observed_log_messages.append(msg)
```

Comment:
Why don't we test the remaining arguments which get logged?

Reply:
The other args are irrelevant. They are the arguments which are merged into msg using the string formatting operator but we are not using it anywhere

Reply:
OK, add this explanation as part of the docstring of `_mock_logging_function`.

Reply:
Then just merge anyway?

```
def _mock_logging_function(msg, *args):
    """Mocks logging.exception."""
    observed_log_messages.append(msg % args)
```

Reply:
We don't follow this pattern anywhere else in the codebase(the one you wrote above). What if there is not a format string specifier in the original log message? Would it work then?

Reply:
That's because this is a standard Python function, so you should be careful to follow the [API it defines](https://docs.python.org/2/library/logging.html#logging.exception).

Reply:
Done!